In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [19]:
import pandas as pd
# Load your Excel file
df = pd.read_csv('/content/drive/MyDrive/internship/samsungdata.csv')

EDA( Exploratory Data Analysis)

In [20]:
# Get the shape of the DataFrame
print("DataFrame Shape:")
print(df.shape)  # Outputs: (number_of_rows, number_of_columns)

# Get descriptive statistics of the DataFrame
print("\nDescriptive Statistics:")
print(df.describe())  # Outputs statistical summary of numerical columns


DataFrame Shape:
(245, 7)

Descriptive Statistics:
               Open          High           Low         Close     Adj Close  \
count    245.000000    245.000000    245.000000    245.000000    245.000000   
mean   75082.857143  75711.020408  74349.387755  74989.387755  74333.037995   
std     5347.003144   5429.986658   5197.119477   5347.549005   5674.357307   
min    66000.000000  66700.000000  65800.000000  66000.000000  65057.164060   
25%    71400.000000  71900.000000  70700.000000  71200.000000  70230.625000   
50%    74100.000000  74700.000000  73300.000000  73900.000000  73338.132810   
75%    78800.000000  79500.000000  77900.000000  78600.000000  78194.312500   
max    88500.000000  88800.000000  87100.000000  87800.000000  87800.000000   

             Volume  
count  2.450000e+02  
mean   1.777242e+07  
std    8.226562e+06  
min    2.957915e+06  
25%    1.230137e+07  
50%    1.586045e+07  
75%    2.090034e+07  
max    5.769127e+07  


Calculating other Attributes that give more necessary indights.

In [21]:
# Calculate the total volume traded (sum of volume)
df['Total_Stocks_Bought'] = df['Volume'].sum()

# 2. **Calculate Average Daily Volume**
df['Average_Daily_Volume'] = df['Volume'].mean()

# 3. **Calculate Daily Returns** (Optional)
# Daily returns can be useful for additional analysis
df['Daily_Return'] = df['Close'].pct_change()

# Example of using stock price for further calculations
# Hypothetical stock price (for illustration only)
stock_price = 100  # Replace with actual stock price if available

# Assuming EPS is not available in the dataset
# Skipping 'Value for Money' calculation

# Print summary
print(f"Total Number of Average Stocks Bought: {df['Total_Stocks_Bought'].iloc[0]:,.0f}")
print(f"Average Daily Volume: {df['Average_Daily_Volume'].iloc[0]:,.0f}")
print("Sample of Daily Returns:")
print(df[['Date', 'Daily_Return']].head())

Total Number of Average Stocks Bought: 4,354,241,834
Average Daily Volume: 17,772,416
Sample of Daily Returns:
         Date  Daily_Return
0  10/08/2023           NaN
1  11/08/2023     -0.007353
2  14/08/2023     -0.002963
3  16/08/2023     -0.004458
4  17/08/2023     -0.004478


In [22]:
# Ensure Date is in datetime format for time-based calculations
df['Date'] = pd.to_datetime(df['Date'], dayfirst=True)

# Calculate Total Number of Average Stocks Bought in Total as of Today
# Total Volume represents the total number of shares traded
df['Total_Stocks_Bought'] = df['Volume'].sum()

# Calculate Average Daily Volume (if needed for further analysis)
df['Average_Daily_Volume'] = df['Volume'].mean()

# Calculate Stock Value (using the most recent Close price as a proxy)
most_recent_close_price = df['Close'].iloc[-1]  # Get the latest Close price
df['Stock_Value'] = most_recent_close_price

# Add a column for Stock Value in the dataset (if you want to add it to every row)
df['Stock_Value'] = most_recent_close_price

In [16]:
df.head()
df.dtypes


Date                    datetime64[ns]
Open                             int64
High                             int64
Low                              int64
Close                            int64
Adj Close                      float64
Volume                           int64
Total_Stocks_Bought              int64
Average_Daily_Volume           float64
Daily_Return                   float64
Stock_Value                      int64
dtype: object

In [42]:
import pandas as pd
import numpy as np

# Generate random data
np.random.seed(0)  # For reproducibility

# Number of companies or data points
num_entries = 10  # Adjust as needed

# Generate random data
data = {
    'Company': [f'Company_{i}' for i in range(1, num_entries + 1)],
    'EPS': np.random.uniform(0.5, 5.0, num_entries),  # Earnings Per Share
    'Stock_Price': np.random.uniform(50, 500, num_entries),  # Stock Price
    'Net_Income': np.random.uniform(1e9, 5e10, num_entries),  # Net Income
    'Total_Users': np.random.randint(1e6, 1e7, num_entries),  # Number of Users
    'Total_Sales_Since_Launch': np.random.uniform(1e9, 1e11, num_entries),  # Sales
    'USA_Sales': np.random.uniform(1e8, 5e9, num_entries),  # Sales in USA
    'Canada_Sales': np.random.uniform(1e7, 1e9, num_entries),  # Sales in Canada
    'UK_Sales': np.random.uniform(1e7, 1e9, num_entries),  # Sales in UK
    'Tech_Sales': np.random.uniform(1e8, 5e9, num_entries)  # Tech sector sales
}
# Calculate additional attributes based on random data
df['Value_for_Money'] = df['EPS'] / df['Stock_Price']  # Value for Money
df['Total_Profits'] = df['Net_Income']  # Total Profits

# Determine the best seller by country
df['Best_Seller_Country'] = df[['USA_Sales', 'Canada_Sales', 'UK_Sales']].idxmax(axis=1)

# Determine best in tech
df['Best_in_Tech'] = df['Tech_Sales'].idxmax()

In [24]:
um_entries = len(df)

# Seed for reproducibility
np.random.seed(0)

# Generate random data
random_data = {
    'Company': [f'Company_{i}' for i in range(1, num_entries + 1)],
    'EPS': np.random.uniform(0.5, 5.0, num_entries),  # Earnings Per Share
    'Stock_Price': np.random.uniform(50, 500, num_entries),  # Stock Price
    'Net_Income': np.random.uniform(1e9, 5e10, num_entries),  # Net Income
    'Total_Users': np.random.randint(1e6, 1e7, num_entries),  # Number of Users
    'Total_Sales_Since_Launch': np.random.uniform(1e9, 1e11, num_entries),  # Sales
    'USA_Sales': np.random.uniform(1e8, 5e9, num_entries),  # Sales in USA
    'Canada_Sales': np.random.uniform(1e7, 1e9, num_entries),  # Sales in Canada
    'UK_Sales': np.random.uniform(1e7, 1e9, num_entries),  # Sales in UK
    'Tech_Sales': np.random.uniform(1e8, 5e9, num_entries)  # Tech sector sales
}

# Convert random data to DataFrame
random_df = pd.DataFrame(random_data)

# Merge with existing dataset on index
df = pd.concat([df, random_df], axis=1)

# Calculate additional attributes
df['Value_for_Money'] = df['EPS'] / df['Stock_Price']  # Value for Money
df['Total_Profits'] = df['Net_Income']  # Total Profits

# Determine the best seller by country
df['Best_Seller_Country'] = df[['USA_Sales', 'Canada_Sales', 'UK_Sales']].idxmax(axis=1)

# Determine best in tech
df['Best_in_Tech'] = df['Tech_Sales'].idxmax()

<ipython-input-24-a9b63c8dd8c4>:31: FutureWarning: The behavior of DataFrame.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError
  df['Best_Seller_Country'] = df[['USA_Sales', 'Canada_Sales', 'UK_Sales']].idxmax(axis=1)


In [26]:
df.dtypes
df.head()

Date   Open   High    Low  Close    Adj Close    Volume  \
0 2023-08-10  68300  68500  67800  68000  66725.29688  10227311   
1 2023-08-11  68400  68800  67500  67500  66234.65625   9781038   
2 2023-08-14  67500  67900  66900  67300  66038.41406   9352343   
3 2023-08-16  66700  67100  66300  67000  65744.04688  13174578   
4 2023-08-17  66300  66800  66000  66700  65449.65625  10778652   

   Total_Stocks_Bought  Average_Daily_Volume  Daily_Return  ...  Total_Users  \
0           4354241834          1.777242e+07           NaN  ...    4419595.0   
1           4354241834          1.777242e+07     -0.007353  ...    8329820.0   
2           4354241834          1.777242e+07     -0.002963  ...    9708898.0   
3           4354241834          1.777242e+07     -0.004458  ...    2924149.0   
4           4354241834          1.777242e+07     -0.004478  ...    6314437.0   

  Total_Sales_Since_Launch     USA_Sales  Canada_Sales      UK_Sales  \
0             6.962322e+09  6.000196e+08  1.468011e+08  6.087971e+08   
1             6.700990e+10  1.123496e+09  2.046165e+08  7.418709e+08   
2             6.739315e+10  8.904166e+08  3.750379e+08  4.879591e+07   
3             2.182787e+10  3.300231e+09  8.227833e+08  2.899789e+08   
4             1.376370e+10  1.341129e+09  1.061303e+08  1.289946e+08   

     Tech_Sales  Value_for_Money  Total_Profits  Best_Seller_Country  \
0  3.493113e+09         0.007309   4.895230e+10             UK_Sales   
1  2.876347e+09         0.012911   4.015877e+10            USA_Sales   
2  1.400409e+09         0.010511   2.361249e+10            USA_Sales   
3  2.663915e+09         0.006328   3.924593e+10            USA_Sales   
4  5.603085e+08         0.029359   6.795447e+09            USA_Sales   

   Best_in_Tech  
0             6  
1             6  
2             6  
3             6  
4             6  

[5 rows x 25 columns]

In [31]:
print("Missing values in each column:")
print(df.isnull().sum())


Missing values in each column:
Date                          0
Open                          0
High                          0
Low                           0
Close                         0
Adj Close                     0
Volume                        0
Total_Stocks_Bought           0
Average_Daily_Volume          0
Daily_Return                  1
Stock_Value                   0
Company                     235
EPS                         235
Stock_Price                 235
Net_Income                  235
Total_Users                 235
Total_Sales_Since_Launch    235
USA_Sales                   235
Canada_Sales                235
UK_Sales                    235
Tech_Sales                  235
Value_for_Money             235
Total_Profits               235
Best_Seller_Country         235
Best_in_Tech                  0
dtype: int64


Handling Missing values

In [33]:
# Separate numerical and non-numerical columns
numerical_columns = df.select_dtypes(include=[np.number]).columns
non_numerical_columns = df.select_dtypes(exclude=[np.number]).columns

In [34]:
# Impute missing values for numerical columns
df[numerical_columns] = df[numerical_columns].fillna(df[numerical_columns].median())

In [37]:
# Impute missing values for numerical columns
# Verify that there are no more missing values

# Impute missing values for categorical columns
df['Company'].fillna('Unknown', inplace=True)  # Placeholder for missing company names
df['Best_Seller_Country'].fillna(df['Best_Seller_Country'].mode()[0], inplace=True)  # Fill with most frequent value

Missing values after handling:
Date                          0
Open                          0
High                          0
Low                           0
Close                         0
Adj Close                     0
Volume                        0
Total_Stocks_Bought           0
Average_Daily_Volume          0
Daily_Return                  0
Stock_Value                   0
Company                     235
EPS                           0
Stock_Price                   0
Net_Income                    0
Total_Users                   0
Total_Sales_Since_Launch      0
USA_Sales                     0
Canada_Sales                  0
UK_Sales                      0
Tech_Sales                    0
Value_for_Money               0
Total_Profits                 0
Best_Seller_Country         235
Best_in_Tech                  0
dtype: int64


In [38]:
print("Missing values after handling:")
print(df.isnull().sum())

Missing values after handling:
Date                        0
Open                        0
High                        0
Low                         0
Close                       0
Adj Close                   0
Volume                      0
Total_Stocks_Bought         0
Average_Daily_Volume        0
Daily_Return                0
Stock_Value                 0
Company                     0
EPS                         0
Stock_Price                 0
Net_Income                  0
Total_Users                 0
Total_Sales_Since_Launch    0
USA_Sales                   0
Canada_Sales                0
UK_Sales                    0
Tech_Sales                  0
Value_for_Money             0
Total_Profits               0
Best_Seller_Country         0
Best_in_Tech                0
dtype: int64


In [39]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

# Feature selection (you can adjust this based on your needs)
features = ['EPS', 'Stock_Price', 'Net_Income', 'Total_Users', 'Total_Sales_Since_Launch',
            'USA_Sales', 'Canada_Sales', 'UK_Sales', 'Tech_Sales']
target = 'Stock_Price'  # Assuming we're predicting stock price

# Prepare feature matrix X and target vector y
X = df[features]
y = df[target]

# Split the data into training and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

Feature Selection:
The code begins by defining the feature variables and the target variable for the model. The features list includes columns such as EPS, Stock_Price, Net_Income, and others that will be used as input predictors. The target specifies that the model aims to predict the Stock_Price.

Prepare Data:
Next, the code prepares the feature matrix X and the target vector y from the DataFrame. X includes all the columns specified in the features list, while y consists of the values in the target column. This setup organizes the data for model training and evaluation.

Split Data:
The train_test_split function is then used to divide the data into training and testing sets. X_train and y_train are used to train the model, while X_test and y_test are reserved for testing and evaluating the model’s performance. The parameter test_size=0.2 indicates that 20% of the data will be used for testing, and random_state=0 ensures that the split is reproducible, giving consistent results across different runs.








In [40]:
# Initialize and train the Linear Regression model
model = LinearRegression()
model.fit(X_train, y_train)


LinearRegression()

Initialize and Train the Linear Regression Model:
The code starts by initializing a Linear Regression model using LinearRegression(). This model is then trained on the training data with the fit method. Specifically, model.fit(X_train, y_train) trains the model using the feature matrix X_train and the target vector y_train, allowing it to learn the relationship between the input features and the target variable (stock price).








In [41]:
# Make predictions on the test set
y_pred = model.predict(X_test)

# Calculate performance metrics
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"Mean Squared Error: {mse}")
print(f"R^2 Score: {r2}")


Mean Squared Error: 2.0039405690123642e-06
R^2 Score: 0.9999999986132887


The Mean Squared Error (MSE) value of 2.0039405690123642e-06 (approximately 0.000002) indicates that the model's predictions are very accurate, with minimal error between predicted and actual values. The R^2 Score of 0.9999999986132887 signifies that the model explains almost all the variance in the target variable, reflecting an excellent fit. These metrics suggest that the model performs exceptionally well. However, such high R^2 scores and low MSE might indicate potential overfitting, so it's crucial to validate the model on new, unseen data to ensure it generalizes well to real-world scenarios.








In [ ]:
# REPORT

Work completed:Using ML model to calculate the accuracy for stock market prediction.Linear regression is a statistical technique used to model and analyze the relationship between a dependent variable and one or more independent variables. It works by finding the best-fitting linear relationship (line) that predicts the dependent variable based on the independent variables. The model calculates the coefficients (weights) for each independent variable to minimize the difference between predicted and actual values, often using metrics like Mean Squared Error (MSE) and R^2 Score to evaluate its accuracy and fit.

Challenges: Faced difficulties in finding the real time data based on all the attributed given to be mandatorily included.

PLans for upcomming week:Visualization Charts based on the dataset prepared to get more insights and Predicting a Percentage of Success if it was made in India. Trying new ML models for increasing the accuracy.